### PR TITLE
Fix check for array kind compatibility

### PR DESCRIPTION
**Pull request by lthls**
This fixes a bug with the `compy` opam package.

### CODE REVIEW

File: middle_end/flambda2/types/provers.ml [hunk 1]
```diff
@@ -630,7 +630,10 @@ let prove_is_array_with_element_kind env t ~element_kind : _ proof =
   | Value Bottom -> Invalid
   | Value (Ok (Array { element_kind = Unknown; _ })) -> Unknown
   | Value (Ok (Array { element_kind = Known element_kind'; _ })) ->
-    Proved (K.With_subkind.equal element_kind' element_kind)
+    Proved
+      (K.equal
```
Comment (Gbury) [inline middle_end/flambda2/types/provers.ml:634]: Maybe here, you could instead make use of the `K.With_subkind.compatible` function ?

Comment (lthls) [reply middle_end/flambda2/types/provers.ml:634]: I started with that, but I'm worried that `compatible` is not symmetrical.
We want to support both cases where the primitive is specialised and the type isn't (like in the example), and the more common case where the primitive is not specialised but the type is.

File: middle_end/flambda2/tests/mlexamples/arrays_compatibility.ml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+(* Example from comby: an array access was improperly simplified to Invalid.
+
+   The issue was that the array is seen as an array of immediates by the
+   primitive, but its type from the environment is an array of values. The two
+   kinds are compatible, so the array access is valid. *)
```
Comment (lukemaurer) [inline middle_end/flambda2/tests/mlexamples/arrays_compatibility.ml:5]: I don't understand. Surely the access _isn't_ in general valid? All we know from the environment is that it's an array of values, which could be an array of blocks.

File: middle_end/flambda2/types/provers.ml (hunk 1 above)
Comment (lukemaurer) [reply middle_end/flambda2/types/provers.ml:634]: Both of these cases could be invalid, though, no? `Array_load (Immediates, _)` on a value array and `Array_set (Values, _)` on an immediate array could both be bad.

File: middle_end/flambda2/tests/mlexamples/arrays_compatibility.ml (hunk 2 above)
Comment (lthls) [reply middle_end/flambda2/tests/mlexamples/arrays_compatibility.ml:5]: It depends on the definition of "valid". Here I mean it as "not provably invalid". The only case were we are allowed to return `Invalid` is when then access is provably invalid (no value of the given array type would be allowed as argument of the primitive).

File: middle_end/flambda2/types/provers.ml (hunk 1 above)
Comment (lthls) [reply middle_end/flambda2/types/provers.ml:634]: `Array_load (Immediates, _)` on a value array returns an immediate (it merely acts as an additional type annotation, just like you could refine a generic type into a specific one using GADTs).
`Array_set (Values, _)` on an immediate array works in the same way: it merely means that in the current context, the primitive can be specialised.

Here are two examples showing those cases:
```ocaml
type _ w = Int : int w | Other : _ w
let load_immediate_on_value_array (type a) (wit : a w) (v : a) =
  let make_array x = [|x|] (* here we get a value array *) in
  let get_int (a : int array) = a.(0) (* here we load with kind Immediates *) in
  let a = make_array v in
  match wit with
  | Int -> get_int a
  | Other -> 0

let set_value_on_immediate_array () =
  let set (a : _ array) x = a.(0) <- x (* Value set *) in
  let a = [|0|] (* Immediate array *) in
  set a 0
```

The example from this PR shows the first case but without GADTs.

File: middle_end/flambda2/tests/mlexamples/arrays_compatibility.ml (hunk 2 above)
Comment (lukemaurer) [reply middle_end/flambda2/tests/mlexamples/arrays_compatibility.ml:5]: Ah, gotcha.

**Comment by lthls** [comment]
I've pushed a better version, that clarifies exactly what we're proving in `prove_is_array_with_element_kind`.

**Comment by lthls** [comment]
And I've added a few extra examples, although some of them are only interesting because of the missing kind for `Values` arrays in Flambda 2.

File: middle_end/flambda2/tests/mlexamples/array_specialisation_examples.ml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+(* Should be simplified to a Values array load, but isn't because of the missing
+   representation for Values arrays in Flambda kinds *)
```
Comment (mshinwell) [inline middle_end/flambda2/tests/mlexamples/array_specialisation_examples.ml:2]: This comment is out of date now, I think?

**Review by mshinwell — CHANGES_REQUESTED** [review]

File: middle_end/flambda2/tests/mlexamples/array_specialisation_examples.ml (hunk 3 above)
Comment (lthls) [reply middle_end/flambda2/tests/mlexamples/array_specialisation_examples.ml:2]: Yes, I'll update the comments in this file (and check that it does get simplified to a Values load).

**Review by mshinwell — APPROVED** [review]

**Review by lukemaurer — APPROVED** [review]

**Comment by mshinwell** [comment]
@chambart Please catch up on this when you can.

**Pull request merged**